sdks/v4-sdk/src/entities/position.ts

```diff
@@ -52,59 +52,262 @@ export class Position {
    * Returns the price of token0 at the lower tick
    */
   public get token0PriceLower(): Price<Currency, Currency> {
-    // TODO: Currency or wrapped token here?
-    throw new Error('Not implemented')
+    return tickToPrice(this.pool.currency0, this.pool.currency1, this.tickLower)
   }
 
   /**
    * Returns the price of token0 at the upper tick
    */
   public get token0PriceUpper(): Price<Currency, Currency> {
-    throw new Error('Not implemented')
+    return tickToPrice(this.pool.currency0, this.pool.currency1, this.tickUpper)
   }
 
   /**
    * Returns the amount of token0 that this position's liquidity could be burned for at the current pool price
    */
   public get amount0(): CurrencyAmount<Currency> {
-    throw new Error('Not implemented')
+    if (this._token0Amount === null) {
+      if (this.pool.tickCurrent < this.tickLower) {
+        this._token0Amount = CurrencyAmount.fromRawAmount(
+          this.pool.currency0,
+          SqrtPriceMath.getAmount0Delta(
+            TickMath.getSqrtRatioAtTick(this.tickLower),
+            TickMath.getSqrtRatioAtTick(this.tickUpper),
```

Comment:
just double checking.. this should be `tickLower` and `tickUpper` if it's less than the `tickLower`?

Reply:
yep! good question.

So on L70 we check where the tickCurrent is. 

If the tickCurrent is below the tickLower  (tickCurrent < tickLower) that means the full range of liquidity is above the current price. I think about it like this in my head:

```

__________|______________|||||||__
          ^              ^     ^
__________tC____________tL_____tU
```

Anything below tC (tickCurrent) is going to be entirely in amount1 and anything above tickCurrent is going to be entirely in amount0. 

Since this function wants to know what the amount0 is, then in this case we need to query for the amount between the tickLower and the tickUpper. Hopefully that makes sense!

Reply:
I don't think I fully understand. If amount0 is everything to the right of `tC`, wouldn't it be between `tC` and `tL`?

Reply:
ah yeah so everything to the right of tickCurrent is ALL of amount0 on the pool, since we are just working with a specific position (who is defined from tL to tU) we just care about the amount0 between tL and tU.

Reply:
if we wanted to know amount1 (where the state of the pool is the same in the above graph.. ie the position is entirely above the current tick) then amount1 = 0 since the position does not touch any of amount1. And you can verify this logic in the amount1() function call